lib/manageiq/cross_repo/runner.rb

```diff
@@ -35,14 +36,22 @@ def run
 
     def run_tests
       with_test_env do
-        system!({"TRAVIS_BUILD_DIR" => test_repo.path.to_s}, "bash", "tools/ci/before_install.sh") if ENV["CI"] && File.exist?("tools/ci/before_install.sh")
-        system!(env_vars, "bin/setup")
-        system!(script_cmd)
+        require "yaml"
+        travis_yml = YAML.load_file(".travis.yml")
+
+        commands =
+          Array(travis_yml["before_install"]) +
+          Array(travis_yml["install"]) +
+          Array(travis_yml["before_script"]) +
+          Array(travis_yml["script"] || script_cmd) +
+          Array(travis_yml["after_script"])
+
+        system!(env_vars, "/bin/bash -c \"#{commands.join("\; ")}\"")
```

Comment:
I had to switch this from running `.each { |command| system!(command) }` to one `system!()` for all the commands in order to share a shell.  Without this the proper version of `node` would be installed by `before_install.sh` but it wasn't available later when `bundle exec rake spec:javascript` was run.

This is also a little closer to what I assume travis does due to the fact that all of our scripts are loaded with `source tools/ci/*.sh` so they have to be loaded into a common shell to have an impact on the test run.

Reply:
The only thing I don't like about this is before, if a command failed, it would stop, but now with a ; separator it runs them all

Reply:
> ... but now with a ; separator it runs them all

Two ideas for this:

1) Swap `;` with `&&`.  A little dirty, but should have the desired effect.
2) Use these commands to build a script file, and use `set -e` (?) to make sure the script fails when running.

The latter takes a bit more code to do, but you probably could then make said script executable, toss it in a template, and then clean up this file a bit.  Ultimately, it would turn into something like this:

```ruby
template = <<-SCRIPT
#!/bin/bash

set -e

<% commands.each do |cmd| -%>
<%= cmd %>
<% end %>
SCRIPT

script = Tempfile.new
script.write ERB.new(template).result(binding)
script.close
FileUtils.chmod("+x", script.path)

system!(script.path)

script.unlink # in an ensure block
```

Reply:
`-e` can also be used without a file, `bash -e -c "..."`, it's orthogonal.

Reply:
Using `system!` with a single string passed to shell, where you have to quote the commands will lead to issues if the commands contain quotes / backslashes...
Can you pass an array of arguments ["/bin/bash", "-c", "..."] to be exec'd as-is?
Other solutions are generating a file with commands, or piping them to bash's stdin.

Reply:
> TODO use awesome spawn to build the command_line

I think @agrare was planning to do this in a followup which would handle the proper quoting of the command line

Reply:
Both `&&` and `set -e` exit silently without explaining where the failure was.

- Worth also using `-x` to have bash print executed commands (prefixed with `+ `).
- Could get more fancy with `trap` e.g. https://disconnected.systems/blog/another-bash-strict-mode/, not sure needed.
- To get really fancy, print Travis magic folding / time measurement markers https://github.com/travis-ci/travis-ci/issues/2285

Reply:
> To get really fancy, print Travis magic folding / time measurement markers travis-ci/travis-ci#2285

Yes!  I was going to suggest this myself!